cmd/cli/version_test.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/openservicemesh/osm/pkg/version"
+)
+
+type fakeRemoteVersion struct {
+	err         error
+	versionInfo *version.Info
+}
+
+func (f *fakeRemoteVersion) proxyGetMeshVersion(pod string, namespace string, clientset kubernetes.Interface) (*version.Info, error) {
+	if f.err != nil {
+		return nil, f.err
+	}
+	return f.versionInfo, nil
+}
+
+func TestGetMeshVersion(t *testing.T) {
+	tests := []struct {
+		name              string
+		namespace         string
+		remoteVersion     *version.Info
+		remoteVersionInfo *remoteVersionInfo
+		controllerPods    []*corev1.Pod
+		err               error
+	}{
+		{
+			name:              "no mesh in namespace",
+			namespace:         "test",
+			remoteVersion:     nil,
+			remoteVersionInfo: nil,
+			controllerPods:    []*corev1.Pod{},
+			err:               errors.Errorf("No mesh found in namespace [%s]", "test"),
+		},
+		{
+			name:              "mesh in namespace and proxyGetMeshVersion fails",
+			namespace:         "test",
+			remoteVersion:     nil,
+			remoteVersionInfo: nil,
+			controllerPods: []*corev1.Pod{
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name:      "controllerPod",
+						Namespace: "test",
+						Labels: map[string]string{
+							constants.AppLabel:               constants.OSMControllerName,
+							constants.OSMAppInstanceLabelKey: "osm",
+						},
+					},
+				},
+			},
+			err: errors.Errorf("Error retrieving mesh version from pod %s in namespace %s)", "controllerPod", settings.Namespace()),
+		},
+		{
+			name:      "mesh in namespace and remote version found",
+			namespace: "test",
+			remoteVersion: &version.Info{
+				Version:   "v0.0.0",
+				GitCommit: "xxxxxxx",
+				BuildDate: "Date",
+			},
+			remoteVersionInfo: &remoteVersionInfo{
+				meshName: "osm",
+				version: &version.Info{
+					Version:   "v0.0.0",
+					GitCommit: "xxxxxxx",
+					BuildDate: "Date",
+				},
+			},
+			controllerPods: []*corev1.Pod{
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name:      "controllerPod",
+						Namespace: "test",
+						Labels: map[string]string{
+							constants.AppLabel:               constants.OSMControllerName,
+							constants.OSMAppInstanceLabelKey: "osm",
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			assert := tassert.New(t)
+
+			buf := bytes.NewBuffer(nil)
+
+			objs := make([]runtime.Object, len(test.controllerPods))
+			for i := range test.controllerPods {
+				objs[i] = test.controllerPods[i]
+			}
+
+			cmd := versionCmd{
+				out:       buf,
+				namespace: test.namespace,
+				clientset: fake.NewSimpleClientset(objs...),
+				remoteVersion: &fakeRemoteVersion{
+					err:         test.err,
+					versionInfo: test.remoteVersion,
+				},
+				remote: true,
+			}
+
+			remoteVersionInfo, err := cmd.getMeshVersion()
+			if err != nil && test.err != nil {
+				assert.Equal(test.err.Error(), err.Error())
+			} else {
+				assert.Equal(test.err, err)
+			}
```

Comment:
Because `test.err` is used as both the error returned by the fake `proxyGetMeshVersion()` and an expected return value from `getMeshVersion()`, the actual contents defined in the test case don't really matter since they'll always match, so checking them for equality doesn't make much sense here. Off the top of my head I'm not coming up with a decent way to refactor this though.

Reply:
I was checking the error contents in case `getControllerPods` failed. I think this would be most helpful when writing the unit tests and the controller pods were created incorrectly. If in the test we expected one error, but a different error was returned due to incorrectly setting up the test we wouldn't know and the expected scenarios wouldn't be tested.

Reply:
In general I would probably define the mocked error returned by `proxyGetMeshVersion()` in the test case separately from the error expected to be returned by `getMeshVersion()`. For the cases currently defined, these would always be the same, so sharing one field for both works.

If you define a test case where `getControllerPods()` is expected to fail, you can set `test.err` to that error you expect, and since `proxyGetMeshVersion()` wouldn't get called then, the test would still pass. But if `getControllerPods()` doesn't fail where you expect it to in that case, then `proxyGetMeshVersion()` will still return that `test.err` and the test will pass, even though it's not failing where you expect it to.

Reply:
Added a `proxyGetMeshVersionErr` field